File: algorithm/leetcodeCpp/strobogrammatic-number.cpp
// 246. Strobogrammatic Number
// Difficulty : Easy

// A strobogrammatic number is a number that looks the same when rotated 180 degrees (looked at upside down).

// Write a function to determine if a number is strobogrammatic. The number is represented as a string.

// For example, the numbers "69", "88", and "818" are all strobogrammatic.

// Time:  O(n)
// Space: O(1)

class Solution {
public:
    bool isStrobogrammatic(string num) {
        const int n = num.size();
        for (int i = 0; i <= n / 2; ++i) {
            if (!lookup.count(num[i]) || lookup[num[i]] != num[n - 1 - i]) {
                return false;
            }
        }
        return true;
    }

private:
    const unordered_map<char, char> lookup{{'0', '0'}, {'1', '1'},
                                           {'6', '9'}, {'8', '8'}, 
                                           {'9', '6'}};
};

File: data-structure-and-algorithm/typescript/leetcode/src/strobogrammatic-number.ts
/*
A strobogrammatic number is a number that looks the same when rotated 180 degrees (looked at upside down).

Write a function to determine if a number is strobogrammatic. The number is represented as a string.

Example 1:

Input:  "69"
Output: true
Example 2:

Input:  "88"
Output: true
Example 3:

Input:  "962"
Output: false
*/

function isStrobogrammatic(num: string): boolean {
  const pairMap = new Map<string, string>([
    ['0', '0'],
    ['1', '1'],
    ['8', '8'],
    ['6', '9'],
    ['9', '6'],
  ]);
  for (let i = 0; i <= num.length / 2; ++i) {
    if (
      !pairMap.has(num[i]) ||
      pairMap.get(num[i]) !== num[num.length - 1 - i]
    ) {
      return false;
    }
  }
  return true;
}
